extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```
@@ -0,0 +1,537 @@
+                            }
+                            newPullRequest.setLongPull(!hasMessagesPending());
+
+                            // notify pull message service to pull messages from brokers.
+                            pullMessageService.putRequest(newPullRequest);
+
+                            // set the earliest pull in case we need to block.
+                            if (null == earliestPullRequest) {
+                                earliestPullRequest = newPullRequest;
+                            }
+                        }
+                    }
+                }
+
+                // Block only when there is no locally pending messages.
+                if (!hasMore && null != earliestPullRequest) {
+                    try {
+                        earliestPullRequest.getCountDownLatch().await();
+                        hasMore = true;
+                    } catch (InterruptedException e) {
+                        LOGGER.error("CountDownLatch await got interrupted", e);
+                    }
+                }
+                return hasMore;
+            }
+
+            @Override
+            public InputRow nextRow() {
+                for (Map.Entry<MessageQueue, ConcurrentSkipListSet<MessageExt>> entry : messageQueueTreeSetMap.entrySet()) {
+                    if (!entry.getValue().isEmpty()) {
+                        MessageExt message = entry.getValue().pollFirst();
+                        InputRow inputRow = theParser.parse(ByteBuffer.wrap(message.getBody()));
+
+                        if (!windows.keySet().contains(entry.getKey())) {
+                            windows.put(entry.getKey(), new ConcurrentSkipListSet<Long>());
+                        }
+                        windows.get(entry.getKey()).add(message.getQueueOffset());
+                        return inputRow;
+                    }
+                }
+
+                // should never happen.
+                return null;
```

can we throw an exception if it does?